### ISSUE TITLE
Prefix match not recognizing subsequent word when it starts with the same letter as the first word

### ISSUE THREAD

**Issue by tobymarsden** [typesense/typesense#348]
## Description

I’m having some trouble getting prefix matches to work as I expect; with a small corpus of 10 documents with names starting with `John`, a prefix search for `John W` correctly shows `John Williams` with a `text_match` higher than the other Johns. But if you start the second word with the same letter that the first word starts with, i.e. `John J`, the text_match is the same for all results and therefore the order isn’t what you’d expect.

## Steps to reproduce

```
curl "http://localhost:8108/collections" \
       -X POST \
       -H "Content-Type: application/json" \
       -H "X-TYPESENSE-API-KEY: xyz" \
       -d '{
         "name": "johns",
         "fields": [
           {"name": "name", "type": "string" }
         ]
       }'

curl "http://localhost:8108/collections/johns/documents/import?action=create" -X POST \
        -H "Content-Type: application/json" \
        -H "X-TYPESENSE-API-KEY: xyz" \
        -d '{ "id": "1", "name": "John Stark" }
        { "id": "1", "name": "John Atwood" }
        { "id": "2", "name": "John Smith" }
        { "id": "3", "name": "John Johnson" }
        { "id": "4", "name": "John Williams" }
        { "id": "5", "name": "John Brown" }
        { "id": "6", "name": "John Jones" }
        { "id": "7", "name": "John Garcia" }
        { "id": "8", "name": "John Miller" }
        { "id": "9", "name": "John Keller" }
        { "id": "10", "name": "John Davis" }'
```

```
curl -H "X-TYPESENSE-API-KEY: xyz" \
"http://localhost:8108/collections/johns/documents/search\
?q=John%20W&query_by=name&per_page=3"

{"facet_counts":[],"found":11,"hits":[{"document":{"id":"4","name":"John Williams"},"highlights":[{"field":"name","matched_tokens":["John","Williams"],"snippet":"<mark>John</mark> <mark>Williams</mark>"}],"text_match":50225924},{"document":{"id":"10","name":"John Davis"},"highlights":[{"field":"name","matched_tokens":["John"],"snippet":"<mark>John</mark> Davis"}],"text_match":33514496},{"document":{"id":"9","name":"John Keller"},"highlights":[{"field":"name","matched_tokens":["John"],"snippet":"<mark>John</mark> Keller"}],"text_match":33514496}],"out_of":11,"page":1,"request_params":{"collection_name":"johns","per_page":3,"q":"John W"},"search_time_ms":8}
```

```
curl -H "X-TYPESENSE-API-KEY: xyz" \
"http://localhost:8108/collections/johns/documents/search\
?q=John%20J&query_by=name&per_page=3"

{"facet_counts":[],"found":11,"hits":[{"document":{"id":"10","name":"John Davis"},"highlights":[{"field":"name","matched_tokens":["John"],"snippet":"<mark>John</mark> Davis"}],"text_match":50226176},{"document":{"id":"9","name":"John Keller"},"highlights":[{"field":"name","matched_tokens":["John"],"snippet":"<mark>John</mark> Keller"}],"text_match":50226176},{"document":{"id":"8","name":"John Miller"},"highlights":[{"field":"name","matched_tokens":["John"],"snippet":"<mark>John</mark> Miller"}],"text_match":50226176}],"out_of":11,"page":1,"request_params":{"collection_name":"johns","per_page":3,"q":"John J"},"search_time_ms":14}
```

## Expected Behavior

The `text_match` for `John Jones` and `John Johnson` should be higher than e.g. `John Davis` and therefore they'd appear first, plus `Jones` and `Johnson` should appear in `matched_tokens`.

## Actual Behavior

The `text_match` for all results is the same and `matched_tokens` only contains `John`, i.e. apparently ignoring the prefix match on the second word.

## Metadata

**Typsense Version**: 0.21.0

**OS**: macOS 11.2.2

**Comment by kishorenc** [comment]
@tobymarsden Thanks. This is an excellent catch. The cause of this issue was that Typesense was searching for `john` once again for the `j` prefix, so ended up counting `john` twice for all tokens. I already have a Docker build where this is fixed:

```typesense/typesense:0.22.0.rcu4```

Can you please verify?

**Comment by jasonbosco** [comment]
Released in v0.22
